### PR TITLE
refactor: cleaning of using package/global defaults

### PR DESCRIPTION
**Pull request by Borda**
While Working on the package, I found the default challenging to follow... I go that package want to have one global setting but as dist it hard to trace and eventually help with refactor by any IDE, so using `dataclass` would be better.

Also, for user, I think using default directly with the `cashier` would be much better, see additional suggestion
```py
def cachier(
    hash_func: _Type_HashFunc = _default_hash_func,
    hash_params: Optional[_Type_HashFunc] = None,
    backend: _Type_Backend = "pickle",
    mongetter: Optional[_Type_Mongetter] = None,
    stale_after: datetime.timedelta = datetime.timedelta.max,
    next_time: bool = False,
    cache_dir: Union[str, os.PathLike] = "~/.cachier/",
    pickle_reload: bool = True,
    separate_files: bool = False,
    wait_for_calc_timeout: int = 0,
    allow_none: bool = False,
):
    ...
```

### CODE REVIEW

File: cachier/core.py [hunk 1]
```diff
@@ -220,12 +173,17 @@ def cachier(
         warn(message, DeprecationWarning, stacklevel=2)
         hash_func = hash_params
     # Override the backend parameter if a mongetter is provided.
```
Comment (shaypal5) [inline cachier/core.py:175]: This comment should be at line 184 after the edits, as it describes the `if callabe(mongetter):` logic.

File: cachier/core.py [hunk 2]
```diff
@@ -234,23 +192,16 @@ def cachier(
             cache_dir=cache_dir,
             separate_files=separate_files,
             wait_for_calc_timeout=wait_for_calc_timeout,
-            default_params=_default_params,
         )
     elif backend == "mongo":
-        if mongetter is None:
```
Comment (shaypal5) [inline cachier/core.py:240]: Are you sure you can delete this check? What if somebody explicitly provides `backend="mongo"` and `mongetter=None`?

File: cachier/cores/base.py [hunk 3]
```diff
@@ -9,6 +9,8 @@
 import abc  # for the _BaseCore abstract base class
 import inspect
 
+from ..config import _Type_HashFunc
```
Comment (shaypal5) [inline cachier/cores/base.py:12]: Is it confusing, perhaps, to import custom types from `config`? What about having a separate `custom_types` module, that both config and other modules can import from?

File: cachier/core.py (hunk 2 above)
Comment (shaypal5) [reply cachier/core.py:240]: Ok I see this was relegated to the mongo core. Nice.

File: cachier/cores/pickle.py [hunk 4]
```diff
@@ -19,19 +19,11 @@
...
 class _PickleCore(_BaseCore):
-    """The pickle core class for cachier.
-
-    Parameters
-    ----------
-    pickle_reload : bool, optional
-        See core.cachier() documentation.
-    cache_dir : str, optional.
-        See core.cachier() documentation.
-
```
Comment (shaypal5) [inline cachier/cores/pickle.py:33]: Why are we dropping parameter documentation of the pickle core? Do we not get, as devs, to have some kind of memory and documentation about these? Isn't separate files, in the very least, unique to the pickle core?

**Review by shaypal5 — CHANGES_REQUESTED** [review]
Ok. This looks good. I really like this PR. See just a couple of small notes I had + getting the tests to pass. And of course, if this requires a README change - make sure to update it (it seems to me this only touches on internal interfaces, not something user-facing, but make sure this is the case).

**Comment by Borda** [comment]
> this requires a README change - make sure to update it (it seems to me this only touches on internal interfaces, not something user-facing, but make sure this is the case)

Correct, I made this change limited so it shall not have any impact on a user

File: cachier/core.py (hunk 2 above)
Comment (Borda) [reply cachier/core.py:240]: It was moved to class not deleted

File: cachier/cores/base.py (hunk 3 above)
Comment (Borda) [reply cachier/cores/base.py:12]: That is fine too

File: cachier/cores/pickle.py (hunk 4 above)
Comment (Borda) [reply cachier/cores/pickle.py:33]: It was not compete so I saw two options drop or remove and as other classes doesn't have it either...

File: cachier/cores/base.py (hunk 3 above)
Comment (Borda) [reply cachier/cores/base.py]: done

**Comment by Borda** [comment]
@shaypal5 any thought why this single test is failing `test_wait_for_calc_applies_dynamically[pickle-None]`?

**Comment by shaypal5** [comment]
No idea. I suggest using git blame to track down the PR, contributor and issue context; this probably tests a very specific issue, and finding the discussion and the code contributed alongside the test should help.

Good luck! :)

(if it's me let me know and I'll delve into my memory vault)

**Comment by Borda** [comment]
@shaypal5 mind checking it again? :flamingo:

**Review by shaypal5 — APPROVED** [review]
Awesome work, man! Thank you for your contribution!

**Pull request merged**